simulator/infra/t/unit_test.cpp

```diff
@@ -208,3 +208,11 @@ TEST_CASE("Find first set")
     auto val = std::make_unique<unsigned>( 0);
     CHECK( find_first_set( *val) == bitwidth<unsigned> );
 }
+
+TEST_CASE("Test uint128 printing")
+{
+     uint128 test_value = narrow_cast<uint128>(UINT64_MAX) + 1;
+     std::ostringstream out;
+     out << test_value;
+     CHECK( out.str() == "18446744073709551616");
+}
```

Comment:
First of all, you may wrap this to a new function:
```c++
std::string uint128_to_string(uint128 number)
{
     std::ostringstream out;
     out << test_value;
     return out.str();
}

TEST_CASE("Test uint128 printing")
{
    CHECK( uint128_to_string(narrow_cast<uint128>(UINT64_MAX) + 1) == "18446744073709551616");
}
```

Then, please add tests for degenerative cases: `0`, `1`, `23` and the following test to check recursive call:
```c++
   CHECK( uint128_to_string(all_ones<uint128>()) == "340282366920938463463374607431768211455")
```